graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/ArrayPrototypeBuiltins.java

```diff
@@ -3456,4 +3476,444 @@ protected Object createGroupResult(JSMapObject map, Map<Object, List<Object>> gr
             return map;
         }
     }
+
+    public abstract static class JSArrayToReversedNode extends JSArrayOperation {
+
+        public JSArrayToReversedNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        protected Object toReversedJSArray(JSArrayObject thisObj) {
+            return toReversed(thisObj);
+        }
+
+        @Specialization(replaces = "toReversedJSArray")
+        protected Object reverseGeneric(Object thisObj) {
```

Comment:
(minor, meaning: don't change it, but maybe learn from my comment;-) )

this (the two specializations where one "replaces" the other) is semantically correct, but will in some cases show suboptimal execution performance. Consider a case where this builtin (at one call-site in the javascript code) is called with a (generic) object once, but then with a JavaScript Array. Then, only the `reverseGeneric` will ever be used, as this matches both cases and "replaced" the more specific `toReversedJSArray` variant.

What you would want here is to have two separate specializations, the `toReversedJSArray` just as it is now, the `reverseGeneric` with no replaces, but a `guard="!isJSArray(thisObj)"`.